presto-main/src/main/java/com/facebook/presto/sql/planner/CachingPlanCanonicalInfoProvider.java

```diff
@@ -68,10 +71,20 @@ public Optional<String> hash(Session session, PlanNode planNode, PlanCanonicaliz
     }
 
     @Override
-    public Optional<List<PlanStatistics>> getInputTableStatistics(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy, boolean cacheOnly)
+    public Optional<List<PlanStatistics>> getInputTableStatistics(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy, boolean callMetaData, boolean cacheOnly)
     {
         CacheKey key = new CacheKey(planNode, strategy);
-        return loadValue(session, key, cacheOnly).map(PlanNodeCanonicalInfo::getInputTableStatistics);
+        Optional<PlanNodeCanonicalInfo> planNodeCanonicalInfo = loadValue(session, key, cacheOnly);
+        if (planNodeCanonicalInfo.isPresent()) {
+            if (planNodeCanonicalInfo.get().getInputTableStatistics().isPresent()) {
+                return planNodeCanonicalInfo.get().getInputTableStatistics();
+            }
+            else if (callMetaData) {
```

Comment:
If callMetaData is true, we will fetch the input table statistics from metastore when not available.